### PR TITLE
Chapter code deprecation

### LINKED ISSUE #93: Deprecate 4-character chapter codes, use chapter member ID everywhere
## Is your feature request related to a problem?
Chapters without a 4-character code cannot be viewed via agagd.usgo.org/chapter/<ID>.

## Describe the feature you'd like to see on the AGAGD.
chapters should be looked up via agagd.usgo.org/chapter/<member ID>. Everywhere we look up or display chapters, we should use the ID (with a lookup for the full chapter name, where relevant) instead of the 4-character chapter code.
Additionally, agagd.usgo.org/member/<chapter ID> should redirect to the chapter page.

## Notes
Discussed in June 2020 dev meeting.

### PR DESCRIPTION
**Pull request by annabunches**
Changes all references to chapter codes to prefer the numerical chapter ID from the Chapters table (for internal references) or the chapter name (for display purposes). Chapter codes still work in URLs for legacy purposes (i.e. to avoid breaking bookmarks) and will print instead of the chapter name if the chapter doesn't have a populated name field.

I was unable to test these changes in dev, since I can't get fixtures to load. The django models seem to be sufficiently out of phase with the DB tables that trying to write via the django app fails in several unpredictable ways. We'll want to do substantial testing on the dev site as a result.

Fixes #93.

### CODE REVIEW

**Comment by EarthSchlange** [comment]
Just considering there is also another fall back in the tables that displays a dash which we might want to change to have a fall back with the chapter codes as well.

https://github.com/usgo/agagd/blob/f15012cb3a16ff6422a1209718639975f1bc0139/agagd/agagd_core/tables.py#L100-L109

File: agagd/agagd_core/views/core.py [hunk 1]
```diff
@@ -201,9 +201,16 @@ def tournament_detail(request, tourn_code):
...
-    chapter = Chapters.objects.get(code=chapter_code)
-    member_table = MemberTable(Member.objects.filter(chapter=chapter_code).order_by('family_name') )
+def chapter_detail(request, chapter_id):
+    try:
+        chapter = Chapters.objects.get(member_id=chapter_id)
+        member_table = MemberTable(Member.objects.filter(chapter_id=chapter_id).order_by('family_name') )
+    except DoesNotExist:
+        # Try the lookup with the 4-letter chapter code. These are deprecated,
+        # but we continue to support them in case users have chapter pages bookmarked.
+        chapter = Chapters.objects.get(code=chapter_id)
```
[the remark this answers is not in the document]
Comment (annabunches) [reply agagd/agagd_core/views/core.py:211]: I believe this *is* looking up the chapter code as previously. The variable name has changed, but it's still just the variable that gets passed in via the router. If we don't find an *id* that matches, we fall back to assuming it is a code (which on the Chapters model is stored as `code`) and trying to look it up that way.

File: agagd/agagd_core/views/core.py [hunk 2]
```diff
@@ -201,9 +201,16 @@ def tournament_detail(request, tourn_code):
...
             }) 
 
-def chapter_detail(request, chapter_code):
-    chapter = Chapters.objects.get(code=chapter_code)
-    member_table = MemberTable(Member.objects.filter(chapter=chapter_code).order_by('family_name') )
+def chapter_detail(request, chapter_id):
+    try:
+        chapter = Chapters.objects.get(member_id=chapter_id)
+        member_table = MemberTable(Member.objects.filter(chapter_id=chapter_id).order_by('family_name') )
+    except DoesNotExist:
```
[the remark this answers is not in the document]
Comment (annabunches) [reply agagd/agagd_core/views/core.py:208]: Yes, though I'll just use `exceptions.DoesNotExist` instead.

**Comment by annabunches** [comment]
re: the tables code, this PR includes changes to one of those methods, looked like I missed one though.

File: agagd/agagd_core/views/core.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (EarthSchlange) [reply agagd/agagd_core/views/core.py:211]: I missed the change of `chapter_code` to `chapter_id` on line 204. Looks good.

**Comment by annabunches** [comment]
That's a really good point. I'll figure out how to do that and update the PR :)

**Comment by annabunches** [comment]
This is now back in a state where I believe it to be done. However, we may want to wait until after merging the fixtures code, so we can do local testing before pushing to the dev server.

**Comment by annabunches** [comment]
OH! There is also another **important blocker** before merging this: some users appear to have `chapter` and `chapter_id` values that do not point to the same club! We need to figure out how many of those cases we have and how to deal with them.

**Comment by vash3g** [comment]
'chapter' is the chapter code and 'chapter_id' is the chapter agaID. This is not properly updated in the membership code to update that column as well.

**Comment by annabunches** [comment]
@vash3g yep, the problem is that several users are now listed in one chapter but in that very chapter's view, are shown as belonging to a different club, and we need to be confident that we know which club they "meant" to be in. (because switching to using all `chapter_id`s will represent a change compared to the data that is currently shown in prod)

**Comment by annabunches** [comment]
The latest changes have now been tested with the new fake fixtures, and appear to work. Please take a look!

**Comment by annabunches** [comment]
Ah, discovered that I didn't test the chapter code redirects, and they're currently broken. Working on it :)

**Comment by annabunches** [comment]
There we go, now fixed :)

**Review by EarthSchlange — APPROVED** [review]
I think this looks fine. Chapter ids seem to be loading correctly. The only thing is if there is an edge case of having duplicate chapter codes. 

image

**Comment by EarthSchlange** [comment]
This should fix the issue of loading `/chapter/[id]` in pull #129.

**Pull request merged**